### PR TITLE
Fix Zones/Change Detection issues with database

### PR DESCRIPTION
**Pull request by jeffbcross**

### CODE REVIEW

**Comment by jeffbcross** [comment]
I split out the unrelated commits into #656

**Comment by jeffbcross** [comment]
@vicb would you mind reviewing this commit? https://github.com/angular/angularfire2/pull/648/commits/2a96fa9521a37d7b519775e058c8b6ed34e47815

**Comment by jeffbcross** [comment]
I removed the unrelated commits and rebased against master now, so down to just the zone fix.

File: src/database/firebase_list_factory.spec.ts [hunk 1]
```diff
@@ -541,6 +541,22 @@ describe('FirebaseListFactory', () => {
       });
     });
 
+    it('should emit values in the correct zone', (done: any) => {
+      Zone.current.fork({
+        name: 'newZone'
+      })
+      .run(() => {
+        let observable = FirebaseListFactory(`${rootDatabaseUrl}/questions`);
```
Comment (vicb) [inline src/database/firebase_list_factory.spec.ts]: const

File: src/database/firebase_list_factory.spec.ts [hunk 2]
```diff
@@ -541,6 +541,22 @@ describe('FirebaseListFactory', () => {
       });
     });
 
+    it('should emit values in the correct zone', (done: any) => {
```
Comment (vicb) [inline src/database/firebase_list_factory.spec.ts]: subscription zone?

File: src/database/firebase_object_factory.spec.ts [hunk 3]
```diff
@@ -136,5 +136,22 @@ describe('FirebaseObjectFactory', () => {
       subscription.unsubscribe();
       expect(firebaseSpy).toHaveBeenCalled();
     });
+
+    it('should emit values in the correct zone', (done: any) => {
```
Comment (vicb) [inline src/database/firebase_object_factory.spec.ts]: same comment

File: src/database/firebase_list_factory.spec.ts [hunk 4]
```diff
@@ -541,6 +541,22 @@ describe('FirebaseListFactory', () => {
...
+      Zone.current.fork({
+        name: 'newZone'
+      })
+      .run(() => {
+        let observable = FirebaseListFactory(`${rootDatabaseUrl}/questions`);
+        ref.remove(() => {
```
Comment (vicb) [inline src/database/firebase_list_factory.spec.ts]: How could the zone be something else `newZone` here ?

I think you should:
- call subscribe from `newZone`,
- trigger the cb from any other zone,
- assert that cb in correct zone

**Comment by vicb** [comment]
Ref to https://github.com/angular/angular/issues/4603

**Comment by jeffbcross** [comment]
@vicb all comments addressed. I pushed a separate commit with the changes so you can see what changed: https://github.com/angular/angularfire2/pull/648/commits/41207e38680c51be91ed1cb9f4d500b71d0eae63

I also updated the commit message with reference to the Angular issue.

File: src/database/firebase_list_factory.spec.ts [hunk 5]
```diff
@@ -541,22 +541,25 @@ describe('FirebaseListFactory', () => {
...
-        let observable = FirebaseListFactory(`${rootDatabaseUrl}/questions`);
-        ref.remove(() => {
-          ref.push('data', () => {
-            subscription = observable.subscribe(data => {
-              expect(Zone.current.name).toBe('newZone');
-              done();
-            });
+        // Creating a new observable so that the current zone is captured.
+        subscription = FirebaseListFactory(`${rootDatabaseUrl}/questions`)
+          .filter(d => d.length)
```
Comment (jeffbcross) [inline src/database/firebase_list_factory.spec.ts]: Actually I'm going to make this check that there is a value of `new-data` to make sure the data is coming from this spec. And I'll use a better value than `new-data`.

File: src/database/firebase_object_factory.spec.ts [hunk 6]
```diff
@@ -137,19 +137,22 @@ describe('FirebaseObjectFactory', () => {
...
-            });
+        // Creating a new observable so that the current zone is captured.
+        subscription = FirebaseObjectFactory(`${rootDatabaseUrl}/questions/${i}`)
+          .filter(d => d.$value === 'new-data')
+          .subscribe(data => {
+            expect(Zone.current.name).toBe('newZone');
+            done();
           });
+
+        expect(Zone.current.name).toBe('ProxyZone');
```
Comment (jeffbcross) [inline src/database/firebase_object_factory.spec.ts]: Oops, needs to be moved out of Zone.run

File: src/database/firebase_object_factory.ts [hunk 7]
```diff
@@ -17,13 +18,13 @@ export function FirebaseObjectFactory (
     isRef: () => ref = <firebase.database.Reference>absoluteUrlOrDbRef
   });
 
-  return new FirebaseObjectObservable((obs: Observer<any>) => {
+  return observeOn.call(new FirebaseObjectObservable((obs: Observer<any>) => {
```
Comment (jeffbcross) [inline src/database/firebase_object_factory.ts]: I think this should match list_factory stylistically, and assign the observable to a const before calling `observeOn`. Makes it more readable.

File: src/database/firebase_list_factory.spec.ts [hunk 8]
```diff
@@ -541,6 +541,27 @@ describe('FirebaseListFactory', () => {
...
+          .filter(d => d
+            .map(v => v.$value)
+            .indexOf('in-the-zone') > -1)
+          .subscribe(data => {
+            expect(Zone.current.name).toBe('newZone');
+            done();
+          });
+      });
+
+      expect(Zone.current.name).toBe('ProxyZone');
```
Comment (vicb) [inline src/database/firebase_list_factory.spec.ts]: remove

File: src/database/firebase_object_factory.spec.ts [hunk 9]
```diff
@@ -136,5 +136,25 @@ describe('FirebaseObjectFactory', () => {
...
+        // Creating a new observable so that the current zone is captured.
+        subscription = FirebaseObjectFactory(`${rootDatabaseUrl}/questions/${i}`)
+          .filter(d => d.$value === 'in-the-zone')
+          .subscribe(data => {
+            expect(Zone.current.name).toBe('newZone');
+            done();
+          });
+      });
+
+      expect(Zone.current.name).toBe('ProxyZone');
```
Comment (vicb) [inline src/database/firebase_object_factory.spec.ts]: remove

File: src/database/firebase_object_factory.ts [hunk 10]
```diff
@@ -26,4 +27,6 @@ export function FirebaseObjectFactory (
 
     return () => ref.off('value', fn);
   }, ref);
+
+  return observeOn.call(objectObservable, new utils.ZoneScheduler(Zone.current));
```
Comment (vicb) [inline src/database/firebase_object_factory.ts]: add TODO: should be in the subscription zone instead

File: src/database/firebase_list_factory.ts [hunk 11]
```diff
@@ -152,7 +153,7 @@ function firebaseListObservable(ref: firebase.database.Reference | firebase.data
       ref.off('child_changed', chgFn);
     }
   });
-  return listObs;
+  return observeOn.call(listObs, new utils.ZoneScheduler(Zone.current));
```
Comment (vicb) [inline src/database/firebase_list_factory.ts]: add TODO: should be in the subscription zone instead

**Comment by vicb** [comment]
LGTM after the comments are addressed

**Comment by jeffbcross** [comment]
Merged at e18da0eb76b2ce48a3f38ac86bd9a27a4038e45b !

**Pull request closed without merging**